ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -99,6 +204,249 @@ public Integer getValue() {
         return currentOperationsCount.get();
       }
     };
-    metricRegistry.register(MetricRegistry.name(NonBlockingRouter.class, "-NumActiveOperations"), activeOperations);
+    metricRegistry.register(MetricRegistry.name(NonBlockingRouter.class, "NumActiveOperations"), activeOperations);
+  }
+
+  /**
+   * Count error based on error type.
+   * <p/>
+   * This method should be called when an {@code Operation} is completed or aborted.
+   * @param exception The exception to be counted.
+   */
+  void countError(Exception exception) {
+    operationErrorRate.mark();
+    if (exception instanceof RouterException) {
+      switch (((RouterException) exception).getErrorCode()) {
+        case AmbryUnavailable:
+          ambryUnavailableErrorCount.inc();
+          break;
+        case InvalidBlobId:
+          invalidBlobIdErrorCount.inc();
+          break;
+        case InvalidPutArgument:
+          invalidPutArgumentErrorCount.inc();
+          break;
+        case OperationTimedOut:
+          operationTimedOutErrorCount.inc();
+          break;
+        case RouterClosed:
+          routerClosedErrorCount.inc();
+          break;
+        case UnexpectedInternalError:
+          unexpectedInternalErrorCount.inc();
+          break;
+        case BlobTooLarge:
+          blobTooLargeErrorCount.inc();
+          break;
+        case BadInputChannel:
+          blobTooLargeErrorCount.inc();
+          break;
+        case InsufficientCapacity:
+          insufficientCapacityErrorCount.inc();
+          break;
+        case BlobDeleted:
+          blobDeletedErrorCount.inc();
+          break;
+        case BlobDoesNotExist:
+          blobDoesNotExistErrorCount.inc();
+          break;
+        case BlobExpired:
+          blobExpiredErrorCount.inc();
+          break;
+        default:
+          unknownErrorCountForOperation.inc();
+          break;
+      }
+    } else {
+      unknownErrorCountForOperation.inc();
+    }
+  }
+
+  /**
+   * Get {@link NodeLevelMetrics} for a given {@link DataNodeId}. If the {@link DataNodeId} as the key does not exist,
+   * a {@link NodeLevelMetrics} corresponding to an unknown {@link DataNodeId} will be returned.
+   * @param dataNodeId The {@link DataNodeId} to be indexed.
+   * @return The {@link NodeLevelMetrics}.
+   */
+  NodeLevelMetrics getDataNodeBasedMetrics(DataNodeId dataNodeId) {
+    return dataNodeToMetrics.containsKey(dataNodeId) ? dataNodeToMetrics.get(dataNodeId) : dataNodeToMetrics.get(null);
+  }
+
+  /**
+   * A metrics class that tracks at the {@link DataNodeId} level. These metrics are collected based on the operation
+   * requests sent to individual {@link DataNodeId}. An operation request is part of an operation, and conveys an actual
+   * request to a {@link com.github.ambry.clustermap.ReplicaId} in a {@link DataNodeId}. An operation request can be
+   * either for a metadata blob, or for a datachunk.
+   */
+  public class NodeLevelMetrics {
```

Comment:
let us revisit this class and keep the bare minimum that we think will be useful.